src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py

```diff
@@ -0,0 +1,410 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import logging
+import os
+import copy
+import json
+import torch
+from schema import And, Optional
+
+from nni.utils import OptimizeMode
+
+from ..compressor import Pruner, LayerInfo
+from ..utils.config_validation import CompressorSchema
+from ..utils.op_dependency import get_layers_no_dependency
+from .pruners import LevelPruner
+from .weight_rank_filter_pruners import L1FilterPruner
+
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+
+
+class NetAdaptPruner(Pruner):
+    """
+    This is a Pytorch implementation of NetAdapt compression algorithm.
+
+    The pruning procedure can be described as follows:
+    While Res_i > Bud:
+        1. Con = Res_i - delta_Res
+        2. for every layer:
+            Choose Num Filters to prune
+            Choose which filter to prunee (l1)
+            Short-term fine tune the pruned model
+        3. Pick the best layer to prune
+    Long-term fine tune
+
+    For the details of this algorithm, please refer to the paper: https://arxiv.org/abs/1804.03230
+    """
+
+    def __init__(self, model, config_list, evaluator, fine_tuner,
+                 optimize_mode='maximize', pruning_mode='channel', pruning_step=0.05, experiment_data_dir='./'):
+        """
+        Parameters
+        ----------
+        model : pytorch model
+            The model to be pruned
+        config_list : list
+            Supported keys:
+                - sparsity : The final sparsity when the compression is done.
+                - op_types : The operation type to prune.
+        evaluator : function
+            function to evaluate the masked model
+        fine_tuner : function
+            function to short-term fine tune the masked model
+        optimize_mode : str
+            optimize mode, 'maximize' or 'minimize', by default 'maximize'
+        pruning_mode : str
+            'channel' or 'fine_grained', by default 'channel'
+        experiment_data_dir : str
+            PATH to save experiment data
+        """
+        # models used for iterative pruning and evaluation
+        self._model_to_prune = copy.deepcopy(model)
+        self._pruning_mode = pruning_mode
+
+        super().__init__(model, config_list)
+
+        self._fine_tuner = fine_tuner
+        self._evaluator = evaluator
+        self._optimize_mode = OptimizeMode(optimize_mode)
+
+        # hyper parameters for NetAdapt algorithm
+        self._pruning_step = pruning_step
+
+        # overall pruning rate
+        self._sparsity = config_list[0]['sparsity']
```

Comment:
We may need an overall configuration section in our config, especially for those algorithms with training schedule.

Reply:
I prefer to refactor this in another PR, together with some related issues, e.g. the calc of overall sparsity.